### ISSUE TITLE
Switch on travis auto-cancellation

### ISSUE THREAD

**Issue by bsipocz** [conda-forge/staged-recipes#2272]
Travis has introduced the auto-cancellation feature a few weeks ago. I think it would be valuable for this repo to switch that on. We use it for astropy and a few related packages for a while, and it indeed saves us some travis queue time which is handy especially at busy times. It can be accessed via the travis settings page and does what says on the tin:

"Auto Cancellation allows you to only run builds for the latest commits in the queue. This setting can be applied to builds for Branch builds and Pull Request builds separately. Builds will only be canceled if they are waiting to run, allowing for any running jobs to finish."

**Comment by bgruening** [comment]
@bsipocz I don't think this is enable by all repositories. I can not see it in bioconda.

**Comment by dopplershift** [comment]
I'm pretty sure the feature is still in beta; you can request to be added to the beta I believe.

**Comment by bsipocz** [comment]
@bgruening - My bad, I assumed, even though being beta, that it became fully public at the same time as the cron jobs (also still beta). However I think any project can opt in, we started it in early December. 
https://groups.google.com/forum/#!topic/astropy-dev/WTWj5YvUs5g

**Comment by jakirkham** [comment]
It goes beyond opting in. We would actually need an API so that we can enable this on all existing feedstocks (now pushing 1700) and any need feedstocks we add.

Not to mention, there is a similar feature for AppVeyor, but we had issues with it. I can't recall what they were ATM, but it may have been something like cancelling builds on branches in addition to outdated PR builds. Behavior like this would be undesirable.

If you can find more information on either of these points, that would be appreciated.

**Comment by bsipocz** [comment]
I remember we had some issues with Appveyor, too (I can't recall the details as it wasn't me who dealt with it), but the travis one seem to work better so far.
Regarding the API, I see why it would be super useful here, but sadly I have no experience with it at all, but it may be useful to ask about it from travis support (support@travis-ci.com), they usually respond quickly and are helpful.

**Comment by jakirkham** [comment]
Given the sheer size of conda-forge, we are way more sensitive to CI issues than other orgs. In some cases, we have fixed or worked around CI issues that others only later realized. Case in point ( https://github.com/scikit-learn/scikit-learn/issues/8058 ). I'm hopeful that this Travis CI feature does work well, but would not be surprised if it also needs tweaking.

Indeed asking Travis CI would be a good idea. Unfortunately I do not have time to explore this near term. Maybe someone else can ask and share what they find.

**Comment by jakirkham** [comment]
This is still a good idea. For now, we have added our own version of this in PR ( https://github.com/conda-forge/staged-recipes/pull/2264 ), which should hold us over in the interim.

**Comment by bsipocz** [comment]
@jakirkham - Yes, your PR seems to have a great impact already, at least the typical travis queue for the conda-forge builds have been down by a few hours since it's merged. However the big disadvantage of it, compared to this Travis CI feature, is that while those builds are cancelled within a minute, they still have to wait in the travis OSX queue for ~1-2 hours to get their ~1 minute of runtime.

In the meantime I think I may have another hack idea to work around this OSX backlog queue. What about adding a linux based job to the top of the matrix (again, problem here is how to do it for all the 1700 feedstocks), and run your script via those. If it passes the OSX jobs are triggered, otherwise all jobs in that build are cancelled. While this is not ideal, this would be something all within the control of ``conda-forge`` (though getting a dedicated linux only thread would be super useful for that, otherwise it's most possible that all the 5 workers are stuck in the OSX queue most of the time).

**Comment by jakirkham** [comment]
Yep, I agree that we can do better.

Unfortunately Linux and OS X builds on Travis CI (at least last I checked), shared the same queue. So one could still be waiting a while for builds to be removed.

We also looked at a service to do the cancellations. However we run into rate limiting issues there.

In any event, lots of discussion and history is under issue ( https://github.com/conda-forge/conda-forge-webservices/issues/79 ). Will close this out so we can consolidate the discussion.
